java/ql/src/semmle/code/java/frameworks/play/Play.qll

```diff
@@ -0,0 +1,155 @@
+ */
+class PlayMvcResultClass extends Class {
+  PlayMvcResultClass() { this.hasQualifiedName("play.mvc", "Result") }
+}
+
+/**
+ * The `play.mvc.Results` class.
+ */
+class PlayMvcResultsClass extends Class {
+  PlayMvcResultsClass() { this.hasQualifiedName("play.mvc", "Results") }
+}
+
+/**
+ * The `play.mvc.Http$RequestHeader` class.
+ */
+class PlayMvcHttpRequestHeader extends RefType {
+  PlayMvcHttpRequestHeader() { this.hasQualifiedName("play.mvc", "Http$RequestHeader") }
+}
+
+/**
+ * A `play.mvc.BodyParser<>$Of` annotation.
+ */
+class PlayBodyParserAnnotation extends Annotation {
+  PlayBodyParserAnnotation() { this.getType().hasQualifiedName("play.mvc", "BodyParser<>$Of") }
+}
+
+/**
+ * A `play.filters.csrf.AddCSRFToken` annotation.
+ */
+class PlayAddCsrfTokenAnnotation extends Annotation {
+  PlayAddCsrfTokenAnnotation() {
+    this.getType().hasQualifiedName("play.filters.csrf", "AddCSRFToken")
+  }
+}
+
+/**
+ * The type `play.libs.F.Promise<Result>`.
+ */
+class PlayAsyncResultPromise extends Member {
+  PlayAsyncResultPromise() {
+    exists(Class c |
+      c.hasQualifiedName("play.libs", "F") and
+      this = c.getAMember() and
+      this.getQualifiedName() = "F.Promise<Result>"
+    )
+  }
+}
+
+/**
+ * The type `java.util.concurrent.CompletionStage<Result>`.
+ */
+class PlayAsyncResultCompletionStage extends Type {
+  PlayAsyncResultCompletionStage() {
+    this.hasName("CompletionStage<Result>") and
+    this.getCompilationUnit().getPackage().hasName("java.util.concurrent")
+  }
+}
+
+/**
+ * The class `play.mvc.Controller` or a class that transitively extends it.
+ */
+class PlayController extends Class {
+  PlayController() {
+    this.extendsOrImplements*(any(Class t | t.hasQualifiedName("play.mvc", "Controller")))
+  }
+}
+
+/**
+ * A Play framework controller action method. These are mappings to route files.
+ *
+ * Sample Route - `POST  /login  @com.company.Application.login()`.
+ *
+ * Example - class gets `index` and `login` as valid action methods.
+ * ```java
+ * public class Application extends Controller {
+ *  public Result index(String username, String password) {
+ *     return ok("It works!");
+ *   }
+ *
+ *   public Result login() {
+ *     return ok("Log me In!");
+ *   }
+ * }
+ * ```
+ */
+class PlayControllerActionMethod extends Method {
+  PlayControllerActionMethod() {
+    this = any(PlayController c).getAMethod() and
+    (
+      this.getReturnType() instanceof PlayAsyncResultPromise or
+      this.getReturnType() instanceof PlayMvcResultClass or
+      this.getReturnType() instanceof PlayAsyncResultCompletionStage
+    )
+  }
+}
+
+/**
+ * The PlayFramework action method parameters, these are a source of user input.
+ *
+ * Example - `username` & `password` are marked as valid parameters.
+ * ```
+ *  public class Application extends Controller {
+ *      public Result index(String username, String password) {
+ *        return ok("It works!");
+ *      }
+ *    }
```

Comment:
```suggestion
 * ```java
 * public class Application extends Controller {
 *   public Result index(String username, String password) {
 *     return ok("It works!");
 *   }
 * }
```